subprojects/ci-mirror/README.md

```diff
@@ -0,0 +1,148 @@
+```bash
+for mirror in $(ls mirrors); do
+    . ./mirrors/${mirror}/config.env
+    ./deploy_acme.sh
+done
+```
+- At least one entitled (SKU MCT2741F3) Node with label `entitled="true"`. This requirement is also for mirrors that don't actually need entitlement, because of OpenShift template limitations (can't conditionally choose whether to add a nodeSelector to a pod in the template).
+```bash
+# Entitle a node, then set its name here:
+oc get nodes
+NODE=<node_name>
+oc label node/$NODE entitled=true
+```
+
+# Cluster Architecture
+Each mirror consists of multiple CRs deployed via an OpenShift template. Each mirror resides in its own namespace.
+These CRs include: (you may need to read further sections in this README to understand all of it)
+- A namespace, to contain all the mirror CRs.
+- A Route + Service to make it publicly accessible. The route is annotated to be managed by openshift-acme in order to generate TLS cert and key
+- A PersistentVolumeClaim to store mirror data
+- A ServiceAccount and SecurityContextConstraints to make sure node entitlements come through to the entitled-yum mirror sync pod. Cluster-scoped, but relevant only for the entitled-yum mirror.
+- A Secret to hold the client-ca certificate in order to authenticate clients
+- A ConfigMap to hold the NGINX configuration file
+- A ConfigMap to hold the script used by the sync container to download the mirror files. This script must be idempotent to avoid re-downloading after a pod restart.
+- A Deployment with a pod template consisting of the the sync containiner as an initContainer and the serve container as a regular container
+
+## Mirror configuration files
+The differences between each mirror are encoded in the `mirrors` subdirectory.
+
+mirrors
+├── .nginx-default.conf # A default NGINX configuration file to be used by the mirrors
+├── dataset
+│   ├── config.env # The parameters for the dataset repo
+│   └── sync.sh # The sync script for the dataset mirror
+└── entitled-yum
+    ├── config.env # The parameters for the entitled-yum repo
+    ├── sync.sh # The sync script for the entitled-yum mirror
+    └── yum.repo.d
+        └── rhods_mirror.repo # An example `yum` repo configuration file to make use of the entitled yum mirror.
+
+## Containers
+Each mirror deployment makes use of two container images:
+- The `sync` init container, hosted in `${QUAY_REPO}:sync`. This container is responsible for downloading all the data required to be served by the mirror
+- The `serve` container, hosted in `${QUAY_REPO}:serve`. This container is responsible for running the NGINX server to serve the contents synced by the init container.
+
+To build & push new versions of the containers images, `podman login` to your quay.io account (which must be authorized to push to the `openshift-psap` organization) and run `./containers/build.sh` to build and `./containers/upload.sh` to push. This must be done for each of the mirrors:
+```bash
+for mirror in $(ls mirrors); do
```

Comment:
could you actually remove the `for` loop?
what that `MIRROR=` must be set from what's available in `$(ls mirrors)`,
it's easier to deploy a single mirror

Reply:
That loop appears 4 times in the README, I would have to change it everywhere and in some of those other places it makes sense to do them all at once.

Actually here specifically I did it with a purpose - the mirror sync takes a very long time, so it better if it happens for both mirrors in parallel. That's why there's a separate loop for waiting for them to finish.

Since this is just an example meant to demonstrate that it should be done for all mirrors, I prefer to leave it like this. Also if anyone wants to deploy just a single mirror - it's trivial to just copy the loop body and define the loop var manually

Reply:
ack, I would personally prefer keeping in the README the steps to deploy *one* mirror, as it's what people would usually do,
but keep it as you like, it's just bikeshedding